keps/sig-cli/0031-datadrivencommands.md

```diff
@@ -0,0 +1,412 @@
+---
+kep-number: 31
+title: Data Driven Commands for Kubectl
+authors:
+  - "@pwittrock"
+owning-sig: sig-cli
+participating-sigs:
+reviewers:
+  - "@soltysh"
+  - "@juanvallejo"
+  - "@seans3 "
+approvers:
+  - "@soltysh"
+editor: TBD
+creation-date: 2018-11-13
+last-updated: 2018-11-13
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# data driven commands
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details](#implementation-details)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Alternatives](#alternatives)
+
+## Summary
+
+Many Kubectl commands make requests to specific Resource endpoints.  The request bodies are populated by flags
+provided by the user.
+
+Examples:
+
+- `create <resource>`
+- `set <field> <resource>`
+- `logs`
+
+Although these commands are compiled into the kubectl binary, their workflow is similar to a form on a webpage and
+could be complete driven by the server providing the client with the request (endpoint + body) and a set of flags
+to populate the request body.
+
+Publishing commands as data from the server addresses cli integration with API extensions as well as client-server
+version skew.
+
+## Motivation
+
+Kubectl provides a number of commands to simplify working with Kubernetes by making requests to
+Resources and SubResources.  These requests are mostly static, with fields filled in by user
+supplied flags.  Today the commands are compiled into the client, which as the following challenges:
+
+- Extension APIs cannot be compiled into the client
+- Version-Skewed clients (old client) may be missing commands for new APIs or send outdated requests
+- Version-Skewed clients (new client) may have commands for APIs that are not present in the server or expose
+  fields not present in older API versions
+
+### Goals
+
+Allow client commands that make a single request to a specific resource and output the result to be data driven
+from the server.
+
+- Address cli support for extension APIs
+- Address user experience for version skewed clients
+
+### Non-Goals
+
+Allow client commands that have complex client-side logic to be data driven.
+
+- Require a TTY
+- Are Agnostic to Specific Resources
```

Comment:
Which ironically was the entire point of subresources and which has mostly failed.

Reply:
I like the idea of subresources too...

Reply:
FWIW, this approach could complement subresources that argument compiled into the client using the same approach to invoke them as it does the Resource APIs.

Reply:
Yes